File: November 2024 Leetcode Solution/November-29.py
from heapq import heappush, heappop
from typing import List

class Solution:
    def minimumTime(self, grid: List[List[int]]) -> int:
        MOVES = [(-1, 0), (0, 1), (1, 0), (0, -1)]
        if grid[0][1] > 1 and grid[1][0] > 1:
            return -1
        rows, cols = len(grid), len(grid[0])
        pq = [(0, 0, 0)]
        seen = [[False] * cols for _ in range(rows)]
        seen[0][0] = True
        while pq:
            time, row, col = heappop(pq)
            for dr, dc in MOVES:
                newRow, newCol = row + dr, col + dc
                if 0 <= newRow < rows and 0 <= newCol < cols and not seen[newRow][newCol]:
                    newTime = time + 1
                    if grid[newRow][newCol] > newTime:
                        newTime += (grid[newRow][newCol] - newTime + 1) // 2 * 2
                    if newRow == rows - 1 and newCol == cols - 1:
                        return newTime
                    seen[newRow][newCol] = True
                    heappush(pq, (newTime, newRow, newCol))
        return -1

File: November 2024 Leetcode Solution/November-29.rs
use std::collections::BinaryHeap;
use std::cmp::Reverse;

impl Solution {
    pub fn minimum_time(grid: Vec<Vec<i32>>) -> i32 {
        let directions = [(-1, 0), (1, 0), (0, -1), (0, 1)];
        let rows = grid.len();
        let cols = grid[0].len();

        if grid[0][1] > 1 && grid[1][0] > 1 {
            return -1;
        }

        let mut visited = vec![vec![false; cols]; rows];
        let mut heap = BinaryHeap::new();
        heap.push(Reverse((0, 0, 0))); 
        visited[0][0] = true;

        while let Some(Reverse((time, row, col))) = heap.pop() {
            for &(dr, dc) in &directions {
                let new_row = row as isize + dr;
                let new_col = col as isize + dc;

                if new_row >= 0
                    && new_row < rows as isize
                    && new_col >= 0
                    && new_col < cols as isize
                    && !visited[new_row as usize][new_col as usize]
                {
                    let mut new_time = time + 1;
                    let (nr, nc) = (new_row as usize, new_col as usize);
                    if grid[nr][nc] > new_time {
                        new_time += (grid[nr][nc] - new_time + 1) / 2 * 2;
                    }

                    if nr == rows - 1 && nc == cols - 1 {
                        return new_time;
                    }

                    visited[nr][nc] = true;
                    heap.push(Reverse((new_time, nr, nc)));
                }
            }
        }

        -1
    }
}
